packages/payment-experience-api/src/api/paytrailCardRedirectSuccessController.ts

```diff
@@ -0,0 +1,121 @@
+import { sendReceiptToCustomer } from '../lib/sendEmail'
+import { getProductAccountingBatch } from '@verkkokauppa/product-backend'
+
+export class PaytrailCardRedirectSuccessController extends AbstractController {
+  protected readonly requestSchema = null
+
+  private static success = (url: URL) => {
+    url.pathname = 'success'
+    return url
+  }
+
+  private static fault = (url: URL) => {
+    url.pathname = 'summary'
+    url.searchParams.append('paymentPaid', 'false')
+    return url
+  }
+
+  protected async implementation(req: Request, res: Response): Promise<any> {
+    const globalRedirectUrl = process.env.REDIRECT_PAYTRAIL_PAYMENT_URL_BASE
+    if (!globalRedirectUrl) {
+      throw new Error('No default paytrail redirect url defined')
+    }
+    let redirectUrl = new URL(globalRedirectUrl)
+    try {
+      const { orderId } = req.params
+
+      if (!orderId) {
+        return res.redirect(
+          302,
+          PaytrailCardRedirectSuccessController.fault(redirectUrl).toString()
+        )
+      }
+
+      const order = await getOrderAdmin({ orderId })
+
+      const serviceConfigurationRedirectUrl = await getPublicServiceConfiguration(
+        {
+          namespace: order.namespace,
+          key: 'ORDER_CREATED_REDIRECT_URL',
+        }
+      )
+
+      if (serviceConfigurationRedirectUrl?.configurationValue) {
+        redirectUrl = new URL(
+          serviceConfigurationRedirectUrl.configurationValue
+        )
+      }
+
+      const payment = await checkPaytrailCardReturnUrl({
+        params: req.query,
+        order,
+      })
+
+      if (payment.status !== 'payment_paid_online') {
+        return res.redirect(
+          302,
+          PaytrailCardRedirectSuccessController.fault(redirectUrl).toString()
+        )
+      }
+
+      logger.info(`Load paytrail product accountings for order ${orderId}`)
+      const productAccountings = await getProductAccountingBatch({
+        productIds: order.items.map((item) => item.productId),
+      })
+
+      logger.info(
+        `Create accounting entry for paytrail order ${orderId} with accountings ${JSON.stringify(
+          productAccountings
+        )}`
+      )
+
+      await createAccountingEntryForOrder({
+        orderId,
+        dtos: order.items.map((item) => {
+          const productAccounting = productAccountings.find(
+            (accountingData) => accountingData.productId === item.productId
+          )
+          if (!productAccounting) {
+            throw new ExperienceError({
+              code: 'failed-to-create-order-accounting-entry',
+              message: `No accounting entry found for product ${item.productId}`,
+              responseStatus: StatusCode.BadRequest,
+              logLevel: 'error',
+            })
+          }
+          return {
+            ...item,
+            ...productAccounting,
+          }
+        }),
+      })
+
+      await sendReceiptToCustomer(payment, orderId, order)
```

Comment:
I was thinking about this, we can send the receipt to customer before we create accounting information because these can be handled manually if needed.